### PR TITLE
CMake refactoring

### LINKED ISSUE #342: --std=c++17 trash
https://github.com/OpenXRay/xray-16/blob/edef4df84555a4d37afa94ded0e99ff64c9ce11e/CMakeLists.txt#L71

Используйте set(CMAKE_CXX_STANDARD 17)

### LINKED ISSUE #414: Warnings in CMAKE setup
/xray-16/bin $ CFLAGS="-march=native" CXXFLAGS="-march=native" cmake ..
-- cotire 1.8.0 loaded.
CMake build Release, CFLAGS=-march=native , CXXFLAGS=-march=native -fpermissive --std=c++17 -Wno-attributes -pipe -Wl,--no-undefined -fvisibility=hidden 
-- Found FreeImage: /usr/include, /usr/lib64/libfreeimage.so
-- Found Intel TBB, includes /usr/include, /usr/lib64/libtbb.so;/usr/lib64/libtbbmalloc.so
CMake Warning (dev) at Externals/luajit.cmake:25 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:28 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:53 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:54 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:55 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:56 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at Externals/luajit.cmake:57 (set):
  implicitly converting 'NUMBER' to 'STRING' type.
Call Stack (most recent call first):
  Externals/CMakeLists.txt:1 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Found readline: /usr/lib64/libreadline.so  
-- Found Curses: /usr/lib64/libcurses.so

### PR DESCRIPTION
**Pull request by Masterkatze**

### CODE REVIEW

File: src/xrParticles/particle_actions_collection.cpp [hunk 1]
```diff
@@ -1622,7 +1622,7 @@ extern void noise3Init();
 #if defined(XR_ARCHITECTURE_X86) || defined(XR_ARCHITECTURE_X64)
 #include <xmmintrin.h>
 #elif defined(XR_ARCHITECTURE_ARM) || defined(XR_ARCHITECTURE_ARM64)
-#include "Externals/sse2neon/sse2neon.h"
+#include "sse2neon.h"
```
Comment (Xottab-DUTY) [inline src/xrParticles/particle_actions_collection.cpp]: Should not be global include, should be included like this:
```suggestion
#include "sse2neon/sse2neon.h"
```

File: src/xrUICore/CMakeLists.txt [hunk 2]
```diff
@@ -1,90 +1,179 @@
...
+    "XML/UIXmlInitBase.cpp"
+    "XML/UIXmlInitBase.h"
+    "XML/xrUIXmlParser.cpp"
+    "XML/xrUIXmlParser.h"
+)
 
-######################################################
+group_sources(SRC_FILES)
 
-source_group("arrow" FILES ${ARROW})
```
Comment (Xottab-DUTY) [inline src/xrUICore/CMakeLists.txt:32]: Source groups removed? This is regression. No way.

Comment (Masterkatze) [reply src/xrUICore/CMakeLists.txt:32]: group_sources(SRC_FILES) does the same thing.

Comment (Xottab-DUTY) [reply src/xrUICore/CMakeLists.txt:32]: Aaah, that's how it works.. Sorry :D

File: cmake/FindOgg.cmake [hunk 3]
```diff
@@ -1,4 +1,3 @@
-#https://raw.githubusercontent.com/freeorion/freeorion/master/cmake/FindOgg.cmake
```
Comment (Xottab-DUTY) [inline cmake/FindOgg.cmake]: Why?

File: Externals/NVTT/src/nvcore/CMakeLists.txt [hunk 4]
```diff
@@ -1,47 +1,62 @@
...
-
-TARGET_LINK_LIBRARIES(nvcore ${LIBS})
-
-INSTALL(TARGETS nvcore
-	RUNTIME DESTINATION bin
-	LIBRARY DESTINATION lib
-	ARCHIVE DESTINATION lib/static)
+project(nvcore)
+
+set(NVCORE_SHARED ON)
```
Comment (Xottab-DUTY) [inline Externals/NVTT/src/nvcore/CMakeLists.txt]: Why? Let's continue to build it as static lib

File: cmake/FindOgg.cmake (hunk 3 above)
Comment (Masterkatze) [reply cmake/FindOgg.cmake]: Typo

File: src/xrCDB/xrCDB.cpp [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 namespace Opcode
 {
-#include "OPCODE/OPC_TreeBuilders.h"
```
Comment (Xottab-DUTY) [inline src/xrCDB/xrCDB.cpp]: Should be OPCODE/OPC_*filename*.h

File: src/Layers/xrAPI/CMakeLists.txt [hunk 6]
```diff
@@ -1,15 +1,30 @@
...
 
+target_compile_definitions(${PROJECT_NAME}
+    PRIVATE
+    -DXRAPI_EXPORTS
+)
+
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    PREFIX ""
+)
+
```
Comment (Xottab-DUTY) [inline src/Layers/xrAPI/CMakeLists.txt:27]: Maybe, use cotire on xrAPI too?
I know it's small, but still..

File: src/Layers/xrRenderPC_GL/CMakeLists.txt [hunk 7]
```diff
@@ -1,70 +1,436 @@
...
+    "../xrRender/ShaderResourceTraits.h"
+    "../xrRender/SH_Atomic.cpp"
+    "../xrRender/SH_Atomic.h"
+    "../xrRender/SH_Constant.cpp"
+    "../xrRender/SH_Constant.h"
+    "../xrRender/SH_Matrix.cpp"
+    "../xrRender/SH_Matrix.h"
+    "../xrRender/SH_RT.h"
+    #"../xrRender/SH_Texture.cpp"
+    #"../xrRender/SH_Texture.h"
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_GL/CMakeLists.txt]: This exact header should be added.
```suggestion
    "../xrRender/SH_Texture.h"
```

File: Externals/luajit.cmake [hunk 8]
```diff
@@ -6,340 +6,347 @@
...
+## LuaJIT Library
+include_directories(${LIB_NAME}
+    PRIVATE
+    ${CMAKE_CURRENT_BINARY_DIR}
+    ${LUAJIT_DIR}
+    ${LUAJIT_DIR}/host
+    ${LUAJIT_DIR}/../dynasm
+)
+
+
```
Comment (Xottab-DUTY) [inline Externals/luajit.cmake]:
```suggestion
```

File: Externals/luajit.cmake [hunk 9]
```diff
@@ -6,340 +6,347 @@
...
+    string(REGEX REPLACE "[^!/]+" ".." LUA_DIR "!/${INSTALL_BIN}/")
+else()
+    # Direct path to installation
+    set(LUA_DIR ${CMAKE_INSTALL_PREFIX} CACHE STRING "Destination from which modules will be resolved. See INSTALL_LMOD and INSTALL_CMOD.")
+endif()
+
+set(LUA_PATH_DEFAULT "./?.lua;${LUA_DIR}${LUA_LDIR}/?.lua;${LUA_DIR}${LUA_LDIR}/?/init.lua;./?/init.lua")
+set(LUA_CPATH_DEFAULT "./?${LUA_MODULE_SUFFIX};${LUA_DIR}${LUA_CDIR}/?${LUA_MODULE_SUFFIX};${LUA_DIR}${LUA_CDIR}/loadall${LUA_MODULE_SUFFIX}")
+
+if(WIN32 AND NOT CYGWIN)
```
Comment (Xottab-DUTY) [inline Externals/luajit.cmake]: There should be a space after all `if`'s and `elseif`'s
```suggestion
if (WIN32 AND NOT CYGWIN)
```

File: src/Layers/xrRenderPC_R1/CMakeLists.txt [hunk 10]
```diff
@@ -1,26 +1,67 @@
...
     ${SDL_INCLUDE_DIRS}
-	)
+)
 
-#list(REMOVE_ITEM ${PROJECT_NAME}__SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/./LevelCompilerLoggerWindow.cpp")
-#list(REMOVE_ITEM ${PROJECT_NAME}__INCLUDES "${CMAKE_CURRENT_SOURCE_DIR}/./LevelCompilerLoggerWindow.hpp")
+target_link_libraries(${PROJECT_NAME}
+    PRIVATE
+    xrCore
+    luabind
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_R1/CMakeLists.txt]:
```suggestion
    xrLuabind
```

File: src/Layers/xrRenderPC_R1/CMakeLists.txt [hunk 11]
```diff
@@ -1,26 +1,67 @@
...
-	)
+)
 
-#list(REMOVE_ITEM ${PROJECT_NAME}__SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/./LevelCompilerLoggerWindow.cpp")
-#list(REMOVE_ITEM ${PROJECT_NAME}__INCLUDES "${CMAKE_CURRENT_SOURCE_DIR}/./LevelCompilerLoggerWindow.hpp")
+target_link_libraries(${PROJECT_NAME}
+    PRIVATE
+    xrCore
+    xrLuabind
+    LuaJIT
```
Comment (Xottab-DUTY) [inline src/Layers/xrRenderPC_R1/CMakeLists.txt]:
```suggestion
    ${LUA_LIBRARIES}
```

**Review by Xottab-DUTY — APPROVED** [review]
Thank you!

File: Externals/cmake/FindReadline.cmake [hunk 12]
```diff
@@ -5,21 +5,21 @@
...
-find_path ( READLINE_INCLUDE_DIR readline/readline.h
-      HINTS ${PC_READLINE_INCLUDEDIR} ${PC_READLINE_INCLUDE_DIRS}
-      PATH_SUFFIXES readline )
+find_path(READLINE_INCLUDE_DIR readline/readline.h
+    HINTS ${PC_READLINE_INCLUDEDIR} ${PC_READLINE_INCLUDE_DIRS}
+    PATH_SUFFIXES Readline)
 
-find_library ( READLINE_LIBRARY NAMES readline
-      HINTS ${PC_READLINE_LIBDIR} ${PC_READLINE_LIBRARY_DIRS} )
+find_library(READLINE_LIBRARY NAMES Readline
```
Comment (iq2luc) [inline Externals/cmake/FindReadline.cmake:16]: It should be **readline** not **Readline**.

**Pull request merged**